src/Swarm/Language/Syntax.hs

```diff
@@ -112,15 +113,19 @@ allDirs = [minBound .. maxBound]
 -- | TODO Information about all directions
 dirInfo :: Direction -> DirInfo
 dirInfo d = case d of
-  Lft -> DirInfo "left" Nothing (\(V2 x y) -> V2 (- y) x)
-  Rgt -> DirInfo "right" Nothing (\(V2 x y) -> V2 y (- x))
-  Back -> DirInfo "back" Nothing (\(V2 x y) -> V2 (- y) (- x))
-  Fwd -> DirInfo "forward" Nothing id
-  North -> DirInfo "north" (Just north) (const north)
-  South -> DirInfo "south" (Just south) (const south)
-  East -> DirInfo "east" (Just east) (const east)
-  West -> DirInfo "west" (Just west) (const west)
-  Down -> DirInfo "down" Nothing (const $ V2 0 0)
+  DLeft -> DirInfo directionSyntax Nothing (\(V2 x y) -> V2 (- y) x)
+  DRight -> DirInfo directionSyntax Nothing (\(V2 x y) -> V2 y (- x))
+  DBack -> DirInfo directionSyntax Nothing (\(V2 x y) -> V2 (- y) (- x))
+  DForward -> DirInfo directionSyntax Nothing id
+  DNorth -> DirInfo directionSyntax (Just north) (const north)
+  DSouth -> DirInfo directionSyntax (Just south) (const south)
+  DEast -> DirInfo directionSyntax (Just east) (const east)
+  DWest -> DirInfo directionSyntax (Just west) (const west)
+  DDown -> DirInfo directionSyntax Nothing (const $ V2 0 0)
```

Comment:
@twitu actually why is `down` not treated as other cardinal directions here? In `toDirection` it is. 😕 

It would be nice to avoid all repetition with expressive helpers, that would just take the necessary info - in this case `west` is all that is different in treatment of other cardinal directions.

Reply:
I have fixed the treatment of `down` and have created a helper function `cardinalDirs :: Map (V2 Int64) Direction` to generate a mapping  for `toDirection`. This makes all the direction related function auto generated from the `dirInfo`.

However I just wanted to check if the results of `cardinalDirs` will be memoized because it's a constant function? Otherwise a new `Map` will be created each time `toDirection` is called.

Reply:
I would guess the Map *will be fully evaluated* once `toDirection` is called - but if you are worried about performance, you can add a benchmark* or use `Map.Strict`.

*) I do not think it will show any difference though

Reply:
By "fully evaluated" do you mean re-calculated? And why so? Why won't the performance difference show up? Sorry if I'm pestering you with this, I'm not sure what to search to understand this.

Reply:
No I meant it should be computed once and then it will be ready to use for every operation after. :+1: 
Since you have given it a name `cardinalDirs` it should be treated that way as it does not depend on input of `toDirection`.
With `Map.Lazy` it can be in a partially evaluated state - say it finds the vector at beginning of the `allDirs` list and then stops. 

In any case the overhead is unlikely to be measurable. :wink: